### PR TITLE
imx6ull: add basic tests for flashnor

### PR DESCRIPTION
**Pull request by jsarzy**
## Description

Add basic test of read/write/erase every sixteenth block of nor flash.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [x] Tested by hand on: imx6ull.

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing linter checks and tests passed.
- [x] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

File: storage/imx6ull-flashnor/flashnor-test.c [hunk 1]
```diff
@@ -0,0 +1,157 @@
+#include <getopt.h>
```
Comment (agkaminski) [inline storage/imx6ull-flashnor/flashnor-test.c:1]: Missing file header

File: storage/imx6ull-flashnor/flashnor-test.c [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+void test_read_write(size_t flashsz, size_t sectorsz)
+{
+	int res = 0;
+	unsigned errors = 0;
+	uint8_t *pattern, *buf;
+
+	printf("%s: ", __func__);
+
+	pattern = malloc(sectorsz);
+	if (!pattern) {
```
Comment (agkaminski) [inline storage/imx6ull-flashnor/flashnor-test.c:34]:
```suggestion
	if (pattern == NULL) {
```

File: storage/imx6ull-flashnor/flashnor-test.c [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+		printf("fail: malloc\n");
+		return;
+	}
+
+	buf = malloc(sectorsz);
+	if (!buf) {
```
Comment (agkaminski) [inline storage/imx6ull-flashnor/flashnor-test.c:40]:
```suggestion
	if (buf == NULL) {
```

File: storage/imx6ull-flashnor/flashnor-test.c [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+			res = -1;
+			break;
+		}
+
+		errors += valcmp(buf, sectorsz, pattern[0]);
+	}
+
+	if (res == 0) {
+		if (errors) {
+			printf("fail: read %u errors\n", errors);
```
Comment (agkaminski) [inline storage/imx6ull-flashnor/flashnor-test.c:75]: Is printf enough for this test? Maybe we would like to return error so we can mark the device as defective and stop any further testing?

File: storage/imx6ull-flashnor/flashnor-test.c [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+void test_erase(size_t flashsz, size_t sectorsz)
+{
+	int res = 0;
+	unsigned errors = 0;
+	uint8_t *buf;
+
+	printf("%s: ", __func__);
+
+	buf = malloc(sectorsz);
+	if (!buf) {
```
Comment (agkaminski) [inline storage/imx6ull-flashnor/flashnor-test.c:96]:
```suggestion
	if (buf == NULL) {
```

File: storage/imx6ull-flashnor/flashnor-test.c [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+	int err, c;
+	meterfs_ctx_t *ctx; /* We use meterfs context because (for now) it's fixed in the driver */
+	storage_t dev = { 0 };
+	unsigned int ecspi_no = 3;
+
+	while ((c = getopt(argc, argv, "e:")) != -1) {
+		switch (c) {
+			case 'e':
+				ecspi_no = strtoul(optarg, NULL, 0);
+				break;
```
[the remark this answers is not in the document]
Comment (agkaminski) [reply storage/imx6ull-flashnor/flashnor-test.c:142]: Or provide `default` case

File: storage/imx6ull-flashnor/flashnor-test.c [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+
+	if ((err = flashnor_ecspiInit(ecspi_no, &dev)) < 0) {
+		printf("fail: failed to initialize device (err: %d)", err);
+		return 1;
+	}
+
+	ctx = (meterfs_ctx_t *)dev.ctx;
+
+	test_read_write(ctx->sz, ctx->sectorsz);
+	test_erase(ctx->sz, ctx->sectorsz);
```
Comment (agkaminski) [inline storage/imx6ull-flashnor/flashnor-test.c:154]: No need for this variable imho
```suggestion
	test_read_write(dev.ctx->sz, ctx->sectorsz);
	test_erase(dev.ctx->sz, ctx->sectorsz);
```

File: storage/imx6ull-flashnor/flashnor-test.c (hunk 4 above)
Comment (jsarzy) [reply storage/imx6ull-flashnor/flashnor-test.c:89]: Now, I stop execution if test fails.

File: storage/imx6ull-flashnor/flashnor-test.c [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+	int (*test[])(size_t, size_t) = {
+		test_read_write,
+		test_erase
+	};
+
+	for (i = 0; i < sizeof(test) / sizeof(test[0]); i++) {
+		// If test fails we abort execution
+		if (test[i](flashsz, sectorsz) < 0)
+			return -1;
+	}
```
Comment (agkaminski) [inline storage/imx6ull-flashnor/flashnor-test.c:163]: What's the point of 2 element function array? No reason to do this complicated way
```suggestion
	if (test_read_write(flashsz, sectorsz) < 0)
		return -1;
	
	if (test_erase(flashsz, sectorsz) < 0)
		return -1;
```

File: storage/imx6ull-flashnor/flashnor-test.c (hunk 5 above)
Comment (agkaminski) [reply storage/imx6ull-flashnor/flashnor-test.c:113]: Missed one

**Review by agkaminski — APPROVED** [review]
LGTM

**Pull request merged**